cmd/gcp-controller-manager/app/node_annotater.go

```diff
@@ -0,0 +1,138 @@
+package app
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	core "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+	compute "google.golang.org/api/compute/v1"
+)
+
+type nodeAnnotater struct {
+	c         clientset.Interface
+	ns        corelisters.NodeLister
+	hasSynced func() bool
+	queue     workqueue.RateLimitingInterface
+	gce       *compute.InstancesService
+}
+
+func newNodeAnnotater(client clientset.Interface, nodeInformer coreinformers.NodeInformer) *nodeAnnotater {
+	annotater := &nodeAnnotater{
+		c:         client,
+		ns:        nodeInformer.Lister(),
+		hasSynced: nodeInformer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 1000*time.Second),
+		), "node-annotater"),
+	}
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    annotater.add,
+		UpdateFunc: annotater.update,
+	})
+	return annotater
+}
+
+func (na *nodeAnnotater) add(obj interface{}) {
+	na.enqueue(obj)
+}
+
+func (na *nodeAnnotater) update(obj interface{}, oldObj interface{}) {
+	node := obj.(*core.Node)
+	oldNode := oldObj.(*core.Node)
+	if node.Status.NodeInfo.BootID != oldNode.Status.NodeInfo.BootID {
+		na.enqueue(obj)
+	}
+}
+
+func (na *nodeAnnotater) enqueue(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	na.queue.Add(key)
+}
+
+func (na *nodeAnnotater) Run(workers int, stopCh <-chan struct{}) {
+	if !controller.WaitForCacheSync("node-annotater", stopCh, na.hasSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(na.work, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (na *nodeAnnotater) processNextWorkItem() bool {
+	key, quit := na.queue.Get()
+	if quit {
+		return false
+	}
+	defer na.queue.Done(key)
+
+	na.sync(key.(string))
+	na.queue.Forget(key)
+
+	return true
+}
+
+func (na *nodeAnnotater) work() {
+	for na.processNextWorkItem() {
+	}
+}
+
+func (na *nodeAnnotater) sync(key string) {
+	node, err := na.ns.Get(key)
+	if err != nil {
+		glog.Errorf("Sync %v failed with: %v", key, err)
+		na.enqueue(key)
```

Comment:
It's probably good for now, but we'll need to figure out how to remove keys that keep failing from the queue

Reply:
Yes, we have the per key crashloop backoff at least right now.